### DISCUSSION TITLE
useRouter() error "Cannot read property 'query' of null"

### DISCUSSION THREAD

**Discussion by fca-dev** [vercel/next.js#11818] [Help]
Within a component, I'm calling useRouter. I want to get the query param (if present) and then pass it on to Query or Mutation functions. Yet, no matter what I do, I get the following error: `"Cannot read property 'query' of null"`. In another application, this is all working (I even am destructing useRouter() to access query immediately - if I do that now, I get `Cannot destructure property `query` of 'undefined' or 'null'.`).

I have no idea what I'm supposed to do! Any help would be greatly appreciated.

```
import React, { useState, useEffect } from "react";
import { useRouter } from "next/router";

const Confirm = () => {
  const router = useRouter();
  console.log("ROUTER", router);
  const query = router.query;
  console.log("query", query);

  useEffect(() => {
 ...do things
  }, [query]);

  return (
    <div id="pdf">
      ...return all the components
    </div>
  );
};

export default Confirm;
```

Next version:    9.3.4

**Comment by mravangi** [comment]
same error here.
Did you find solution?

**Reply by adhorodyski** [reply]
Bumping, same here. For me **jest** had multiple errors with desctructured query object:

`const { query } = useRouter()`

**Most of them** are now gone by just not destructuring it:

`const router = useRouter()` and then using query on the router object.

Fun Fact - app runs completely fine, it's just **jest** that fails occasionally - I can't however eliminate all of the errors.

**Comment by sin-to-jin** [comment]
it is basically the same situation with jest.

**Reply by sin-to-jin** [reply]
Currently, all routers are `spyOn` for testing.

```tsx
const router = //...
jest.spyOn(hooks, 'useRouter').mockImplementation(() => (router));
```

**Comment by ilhamwahabi** [comment]
This is how I did it

```ts
const { query } = useRouter() || { query: { text: '' } }
```

In my case I need to get `text` from query parameter, so If i'm not adding that it will yell error

**Comment by luciomartinez** [comment]
Following @ogasawaraShinnosuke suggestion, I mocked the method as follows.

```
jest.mock('next/router', () => ({
  useRouter: () => ({
    query: { myProp: 'myValue' },
  }),
}));
```

**Comment by itxnaeem007** [comment]
Implement this solution. and you have to do this inside **page** folder.

```typescript
import React from 'react'
import Swap from '../views/Swap'

const SwapPage = ({query}) => {
   return <Swap query={query} />
}
SwapPage.getInitialProps = async ({ query }) => {
  return { query }
}
export default SwapPage
```
